src/hazelcore/dynamics/EvaluatorStep.re

```diff
@@ -0,0 +1,514 @@
+open Sexplib.Std;
+
+module EvalCtx = {
+  [@deriving sexp]
+  type t =
+    | Mark
+    | Ap1(t, DHExp.t)
+    | Ap2(DHExp.t, t)
+    | BinBoolOp1(DHExp.BinBoolOp.t, t, DHExp.t)
+    | BinBoolOp2(DHExp.BinBoolOp.t, DHExp.t, t)
+    | BinIntOp1(DHExp.BinIntOp.t, t, DHExp.t)
+    | BinIntOp2(DHExp.BinIntOp.t, DHExp.t, t)
+    | BinFloatOp1(DHExp.BinFloatOp.t, t, DHExp.t)
+    | BinFloatOp2(DHExp.BinFloatOp.t, DHExp.t, t)
+    | Cons1(t, DHExp.t)
+    | Cons2(DHExp.t, t)
+    | Pair1(t, DHExp.t)
+    | Pair2(DHExp.t, t)
+    | Let(DHPat.t, t, DHExp.t)
+    | Inj(HTyp.t, InjSide.t, t)
+    | NonEmptyHole(
+        ErrStatus.HoleReason.t,
+        MetaVar.t,
+        MetaVarInst.t,
+        VarMap.t_(DHExp.t),
+        t,
+      )
+    | Cast(t, HTyp.t, HTyp.t)
+    | FailedCast(t, HTyp.t, HTyp.t)
+    | InvalidOperation(t, InvalidOperationError.t)
+    | ConsistentCase(t, list(DHExp.rule), int)
+    | InconsistentBranches(
+        MetaVar.t,
+        MetaVarInst.t,
+        VarMap.t_(DHExp.t),
+        t,
+        list(DHExp.rule),
+        int,
+      );
+};
+
+let rec is_boxedval = (d: DHExp.t): bool =>
```

Comment:
you don't need this and is_indet to be defined by case analysis anymore since the step function returns whether it is indet, boxedval or steppable.

Reply:
I wrote a totally new code in EvaluatorStep2.re. It didn't contain any function like is_indet.

Reply:
oh sorry, missed that!